src/catalog/database_catalog.cpp

```diff
@@ -3,35 +3,419 @@
+  *oid_entry = ns;
+  std::vector<storage::TupleSlot> index_results;
+  namespaces_oid_index_->ScanKey(*txn, *pr, &index_results);
+  if (index_results.empty()) {
+    delete[] buffer;
+    return false;
+  }
+  TERRIER_ASSERT(index_results.size() == 1, "Namespace OID not unique in index");
+
+  // Step 2: Select from the table to get the name
+  pr = table_pri.InitializeRow(buffer);
+  if (!namespaces_->Select(txn, index_results[0], pr)) {
+    // Nothing visible
+    delete[] buffer;
+    return false;
+  }
+  auto name_varlen = *reinterpret_cast<storage::VarlenEntry *>(pr->AccessForceNotNull(table_pm[NSPNAME_COL_OID]));
+
+  // Step 3: Delete from table
+  if (!namespaces_->Delete(txn, index_results[0])) {
+    // Someone else has a write-lock
+    delete[] buffer;
+    return false;
+  }
+
+  // Step 4: Delete from oid index
+  pr = oid_pri.InitializeRow(buffer);
+  oid_entry = reinterpret_cast<namespace_oid_t *>(pr->AccessForceNotNull(table_pm[NSPOID_COL_OID]));
+  *oid_entry = ns;
+  namespaces_oid_index_->Delete(txn, *pr, index_results[0]);
+
+  // Step 5: Delete from name index
+  pr = name_pri.InitializeRow(buffer);
+  auto name_entry = reinterpret_cast<storage::VarlenEntry *>(pr->AccessForceNotNull(table_pm[NSPNAME_COL_OID]));
+  *name_entry = name_varlen;
+  namespaces_name_index_->Delete(txn, *pr, index_results[0]);
+
+  // Finish
+  delete[] buffer;
+  return true;
+}
+
+namespace_oid_t DatabaseCatalog::GetNamespaceOid(transaction::TransactionContext *txn, const std::string &name) {
+  // Step 1: Read the name index
+  std::vector<col_oid_t> table_oids{NSPNAME_COL_OID};
+  // NOLINTNEXTLINE
+  auto [table_pri, table_pm] = namespaces_->InitializerForProjectedRow(table_oids);
+  auto name_pri = namespaces_name_index_->GetProjectedRowInitializer();
+  // Buffer is large enough to hold all prs
+  byte *buffer = common::AllocationUtil::AllocateAligned(table_pri.ProjectedRowSize());
+  auto pr = name_pri.InitializeRow(buffer);
+  // Scan the name index
+  auto *name_entry = reinterpret_cast<storage::VarlenEntry *>(pr->AccessForceNotNull(0));
+  *name_entry = postgres::AttributeHelper::CreateVarlen(name);
+  std::vector<storage::TupleSlot> index_results;
+  namespaces_name_index_->ScanKey(*txn, *pr, &index_results);
+  if (index_results.empty()) {
+    delete[] buffer;
+    return INVALID_NAMESPACE_OID;
+  }
+  TERRIER_ASSERT(index_results.size() == 1, "Namespace name not unique in index");
+
+  // Step 2: Scan the table to get the oid
+  pr = table_pri.InitializeRow(buffer);
+  if (!namespaces_->Select(txn, index_results[0], pr)) {
+    // Nothing visible
+    delete[] buffer;
+    return INVALID_NAMESPACE_OID;
+  }
+  auto ns_oid = *reinterpret_cast<namespace_oid_t *>(pr->AccessForceNotNull(table_pm[NSPOID_COL_OID]));
+  delete[] buffer;
+  return ns_oid;
+}
+
+template <typename Column>
+bool DatabaseCatalog::CreateTableAttribute(transaction::TransactionContext *txn, uint32_t class_oid, const Column &col,
```

Comment:
This is templated but is named "CreateTableAttribute".  Should this just be "CreateAttribute" because its valid for both tables and indexes?  If so, the 'default_val' parameter should probably be renamed to something more generic like "stored_expression".